internal/states/statefile/version4.go

```diff
@@ -710,14 +767,15 @@ type resourceStateV4 struct {
 	Type           string                  `json:"type"`
 	Name           string                  `json:"name"`
 	EachMode       string                  `json:"each,omitempty"`
-	ProviderConfig string                  `json:"provider"`
+	ProviderConfig string                  `json:"provider,omitempty"`
 	Instances      []instanceObjectStateV4 `json:"instances"`
 }
 
 type instanceObjectStateV4 struct {
-	IndexKey interface{} `json:"index_key,omitempty"`
-	Status   string      `json:"status,omitempty"`
-	Deposed  string      `json:"deposed,omitempty"`
+	IndexKey       interface{} `json:"index_key,omitempty"`
+	Status         string      `json:"status,omitempty"`
+	Deposed        string      `json:"deposed,omitempty"`
+	ProviderConfig string      `json:"provider,omitempty"`
```

Comment:
```suggestion
	ProviderInstance string      `json:"provider,omitempty"`
```

This is totally a nit, but I think it's best to try to be as consistent as we can about provider config vs. provider instance in _new_ code, even if we still have a bunch of followup work to do to make the _old_ code more consistent about it.

Reply:
IMO this is a bit confusing since I expect resource- and instance- level provider fields to be the same logically. They could contain the same value - the only difference is that the instance level is specific to this particular instance. To me, their logical meaning is the same and thus they should be both named the same way.

I would keep it as a `ProviderConfig`, especially if we change `writeStateV4` to write resource level provider if all the instances have the same provider reference (in this case, they both are fully identical and can be keyed).

PS I understand the resource level provider field can't contain keyed references, but IMO we should not reflect this behavior in the naming.

Reply:
I think the main thing I'd like to avoid is any case where we're using the terminology "provider configuration" (or an abbreviation thereof) to refer to what is actually a "provider instance", since the distinction is already quite subtle even if we _do_ name things consistently.

Perhaps instead we should change both the instance-level and the resource-level field to be called `ProviderInstance`, so that we can both use the correct terminology _and_ be internally-consistent in this code? A provider instance address for a single-instance provider configuration is syntactically identical to that provider configuration's address anyway, so there isn't really anything wrong with calling the resource-level address also a "provider instance address", even if for now we know that we'll only ever be storing instances of provider configurations that don't have instance keys.

Reply:
The option to rename both resource- and instance- level provider fields to `ProviderInstance` looks good to me.

Reply:
My preference is to leave this as it is for now and revisit when we take a deeper look at provider address naming